deprecatedError + restricted annotation

eed3si9n opened the pull request:
This implements `restricted` annotation as a generic form of `deprecated` annotation. `restricted` takes `label` and `defaultSeverity` parameter.

The user can configure how the usages of restricted API are handled by passing in `-Wconfig` config expressions.

A warning config expression may be `<label>`, `<label>:<severity>`, or `<label>:<prefix>:<severity>`

`<label>` is a String passed into `@restricted`.
`<severity>` must be one of "error", "warning", "info", or "none": default: warning.
`<prefix>` denotes the prefix of the restricted API. For example, `"foo.*"` will denote
that this config would apply only to calls under foo package/object.
Note: "*" is simply ignored.

The following configuration would escalate apiMayChange restriction to an error:
```
-Wconfig apiMayChange:foo.*:error
```

dwijnand commented:
This is great!

Just a point of admin: this isn't backwards-compatible with 2.13.0, so this would have to land in either 2.13.0-RC1 or slip to 2.14.

som-snytt commented:
There's a WIP of Adriaan's [deprescalator](https://github.com/scala/scala/pull/7728). It makes deprecations go up, down, or silent, by use site or deprecatee.

I haven't fully digested the feature requests on the ticket for ad-hoc `@deprecation`, or maybe it's ad-hoc `@migration`.

nafg commented:
Isn't it redundant with `@compileTimeOnly`? I guess this one wouldn't let a macro remove it, but is that necessary?

eed3si9n commented:
@nafg `@compileTimeError` and `@compileTimeOnly` are similar from the point of view of compiler implementation, but the intents are different. The "only" in compileTimeOnly says it admits the method up until type checking, but compileTimeError is saying no the whole time.
IMO it's useful to have this as a separate construct, so for example tools can potentially do special things (not include in Scaladocs, tab completion etc) with the restligeist methods.

som-snytt commented:
I've got a PR almost ready for `@compileTimeEspecially`, for type params which we especially expect to confer compile-time knowledge and benefits. It will also bump `-Winfer-any`, like `-Werror:infer-any`.
```
def f[@compileTimeEspecially("I better know what I'm doing") A](x: Any) =
  x match { case _: A => }
```
OK, that started out as a joke.

dwijnand commented on src/library/scala/collection/mutable/AnyRefMap.scala:
```diff
@@ -293,8 +293,8 @@ class AnyRefMap[K <: AnyRef, V] private[collection] (defaultEntry: K => V, initi
-  /** Adds a new key/value pair to this map and returns the map. */
-  def +=(key: K, value: V): this.type = { update(key, value); this }
+  @compileTimeError("use += with a tuple argument instead")
+  def +=(key: K, value: V): this.type = ???
```
Should we lose this method?  The replacement requires a tuple allocation.

eed3si9n replied:
You can migrate to `addOne(K, V)` if you want to avoid that.

dwijnand replied:
Makes me thing of BigDecimal "add" vs Int +, but ok.  Good to know.  Thanks.

adriaanm commented:
I would like to generalize the name to subsume `internalOnly`, `apiMayChange`, `deprecated`, `poorPerformance`,....

`restricted(message: String, label: String, severity: Int)`

we could configure according to label and severity for a myriad ways of telling the user off

dwijnand commented:
That assumes those all naturally fit in an int axis, which I don't this is true.  I think it's plausible to imagine some users wanting to see those `poorPerformance` warnings but others being fine with the `apiMayChange` ones, and others wanting the inverse.

What's your driver for trying to unify?

adriaanm commented:
I don’t want to keep adding annotations for other axis. That’s an order of
magnitude less scalable than the concern with the int for the severity. It
could be an enum. Or a string. In any case, the confit would be adhoc to
the project and configurable.

On Mon, 11 Mar 2019 at 18:15, Dale Wijnand <notifications@github.com> wrote:

> That assumes those all naturally fit in an int axis, which I don't this is
> true. I think it's plausible to imagine some users wanting to see those
> poorPerformance warnings but others being fine with the apiMayChange
> ones, and others wanting the inverse.
>
> What's your driver for trying to unify?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/7790#issuecomment-471634950>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjyyEighIg7NtZ6WGMP54wGphwwQOsks5vVo9KgaJpZM4bOQAU>
> .
>

dwijnand commented:
Oh right, so it would be all user-side.  In that case, it would lose value from having shared, universal semantics.

Would the enum be project-side too, or in the standard library?  If it's project-side, then it would lose shared, universal semantic value (such as the `apiMayChange` + MiMa use-case).  If it's library-side, then I'd be fine with it.

som-snytt commented:
`conditionalWarning`.

eed3si9n commented:
So if we reverse engineer the meaning of `@restricted(...)` that would be:
- advisory about the usage of the method
- sometimes about lifecycle of the method (incubating, deprecated, gone)
- that could manifest itself as compilation warnings or error

IIUC, the advisory category (`apiMayChange`, `deprecation`) would be `label: String`. If we want the tool to take special for the context, it does make sense to make label an enum like @dwijnand is saying. With String, we can envision irregularities such as `"API May Change"` vs `"ApiMayChange"` vs `"APIMayChange"`..

`severity: Int` should be able to denote forcefulness of the advisory, like `Severity.AlwaysError`, `Severity.AlwaysWarn`, `Severity.Default` etc.

- My `@compileTimeError` annotation becomes some category (`ApiChange`?) with `Severity.AlwaysError`.
- `@deprecated` annotation becomes `Deprecation` category with `Severity.Default` which gets promoted to a warning under `-deprecation`.

If we imagine `-conditionalWarning:<advisory category>` flag, `-deprecation` might desugar to `-conditionalWarning:Deprecation`.

som-snytt commented:
I'll have to come up with a different name for the `deprescalator`. Or just make it a retronym for how depressed you are at the end of a compilation run.

adriaanm commented:
The annotation and the mechanism to configure whether and how severely to warn should be in the stdlib & compiler, the rest should be configurable.

We could have a few default behaviours, like warning on using an internal member from another package.

adriaanm commented:
I can't extemporaneously come up with a better name, but let's first focus on the semantics. I think we're getting somewhere!

dwijnand commented:
To avoid the issues caused by adding variants to an enum / sealed hierarchy (around compile-time match exhaustion and runtime matching errors) maybe best have it be unsealed so consumers would have to handle the default case.  Optionally with a (scala only) package private constructor, depending on if it should also be user-extendible or not.

> * My `@compileTimeError` annotation becomes some category (`ApiChange`?)

Wouldn't it be in the `Gone` category?

And for that later 👨‍🎨 naming session:

> * `@deprecated` annotation becomes `Deprecation` category with `Severity.Default` which gets promoted to a warning under `-deprecation`

Perhaps `Allow`, `Pass`, or `Off` (or, for pun points, `Severity.Not` 🙂)

Also, I think it would be good if the in-code names for the categories matched the command-line names.  So if it's `Deprecated` in code then it should be `Deprecated` as a command-line argument.  No needs-to-be-learnt string conversion functions #sbt-life-lessons

Finally, for the ultimate 👨‍🎨: `@lint` instead of `@restricted`?

adriaanm commented:
I definitely want the category to be user-extensible/definable, which is why I suggest to let it be a String. What good is the type system if we really just want users to be able to come up with sensible names and provide them on the command line -- unless we port the type checker to bash, we won't get much benefit from static checks here.

The severity is likely a bit more closed world, so we could do an enum there.

I like `@lint`!

dwijnand commented:
> I definitely want the category to be user-extensible/definable, which is why I suggest to let it be a String. What good is the type system if we really just want users to be able to come up with sensible names and provide them on the command line -- unless we port the type checker to bash, we won't get much benefit from static checks here.

We'd be using the type system in the tools that consume those categories (like MiMa + ApiMayChange), which would provide a little more structure than "please, community, use this string here". So, perhaps:
```scala
class Category private[annotation]

object Category {
  final case object Deprecated extends Category
  final case object ApiMayChange extends Category
  final case class Custom(identifier: String) extends Category {
    override def toString = s"Custom($identifier)"
  }
}
```

adriaanm commented:
I see little benefit to that. Some, but not enough to warrant the types.

On Tue, 12 Mar 2019 at 10:39, Dale Wijnand <notifications@github.com> wrote:

> I definitely want the category to be user-extensible/definable, which is
> why I suggest to let it be a String. What good is the type system if we
> really just want users to be able to come up with sensible names and
> provide them on the command line -- unless we port the type checker to
> bash, we won't get much benefit from static checks here.
>
> We'd be using the type system in the tools that consume those categories
> (like MiMa + ApiMayChange), which would provide a little more structure
> than "please, community, use this string here". So, perhaps:
>
> class Category private[annotation]
> object Category {
>   final case object Deprecated extends Category
>   final case object ApiMayChange extends Category
>   final case class Custom(identifier: String) extends Category {
>     override def toString = s"Custom($identifier)"
>   }
> }
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/7790#issuecomment-471926129>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjyw_Tmps1OPvLzfDygSjDYLP3tEUVks5vV3XvgaJpZM4bOQAU>
> .
>

dwijnand commented:
String literals being so readily available I fear we'd immediately have both "deprecated" and "deprecation" as categories, because some users didn't see the `Category.Deprecated` constant.

adriaanm commented:
I don't think that's a big deal. This is an optional, cultural thing that will mostly be driven from outside configuration outside the purview of the type checker. I would rather have zero friction to free choice of categories. It's not about correctness/exhaustivity but rather about being helpful. Having a few built-in types and then a `Custom` "cop-out" does not feel like a substantial UX gain to me. 

If they want the benefits of the few built-in categories we'll eventually support, they'll have to read the docs and fix their spelling.

dwijnand commented:
OK.  And I guess we can attach documentation to those constants, detailing the intended semantics and usage of the category.

eed3si9n commented:
I talked to @szeiger about the removal of the multi-param `+` and he said they are intentionally kept there with `@deprecated` for cross building purpose in 2.13, so I'll drop the commit.

`@compileTimeError` / `@lint` will be useful beyond that hopefully.

som-snytt commented:
I suspect that the generalization of deprecation to `(label, message, since)` is sufficient.

The "severity" should work as for deprecation, namely, a "conditional warning" that can be selectively boosted or suppressed by `-deprecation-policy`, which would be generalized to include the label. (The command line arg could be renamed if there is a better name.)

The deprescalator model would allow selecting by label, use site, definition site, and version. The policy is entirely externalized.

The "severity" axis just complicates the picture. It's a warning! You can turn warnings into errors! You can suppress warnings! Then you're done.

som-snytt commented:
@eed3si9n I think this is a good example of why I shouldn't have to decide if it's an "error" at the definition site. It's just a deprecation (for some reason) and policy determines if I want to fail my build.

eed3si9n commented:
@som-snytt 
> The "severity" axis just complicates the picture. It's a warning! You can turn warnings into errors! You can suppress warnings! Then you're done.

The flexibility of label policy might collide with out-of-box user experience in case of my original use case:

```scala
  @compileTimeError("use ++ with an explicit collection argument instead")
  def + (elem1: A, elem2: A, elems: A*): C = ???
```

The implementation is `???`, so there shouldn't be any opt-in or -out of this being a compilation error. I want the library authors to be able to specify that instead of pushing that to library users. The current alternative is removal of the method.

soronpo commented:
> How about `restricted(message: String, label: String, severity: Int)`?

Perhaps a dumb question, but shouldn't the signature have `Singleton` annotations?
`restricted(message: String with Singleton, label: String with Singleton, severity: Int with Singleton)`

eed3si9n commented:
> Finally, for the ultimate 👨‍🎨: `@lint` instead of `@restricted?`

`@lint` is better than `@restricted` for what's going on, but it still over-claims the noun space. Linting could vary from enforcing styles to detecting `Any` inference etc.

How about `@apiStatus`?

```scala
@apiStatus(ApiStatus.apiMayChange, message = "API might change", since = "2.5.21", severity = Severity.none)
object StreamRefs {
... 
}

@apiStatus(ApiStatus.internalOnly, since = "2.5.21", severity = Severity.none)
object Milli {
... 
}
```

To omit severity out of this mix, as suggested by @som-snytt, maybe we add 2 annotations `@apiStatus` and `@apiStatusError` both with `(label: String, message: String, since: String)`?

som-snytt commented:
I'm trying out `compileTimeError` as a specialized `deprecated` that is escalated to error. Maybe there is a better name for it. Since deprecated is already about API, I would just stick with deprecated as a general concept. Elements are deprecated for many reasons, which is indicated by the message, but for convenience of sorting, filtering, escalating we add a label that nominally categorizes.

A "hard deprecation" indicated by `compileTimeError` might have an implementation, but it's so terrible that ordinary deprecation is insufficient. Can I de-escalate it? Or maybe it has no implementation but I don't exercise it, I just want my project to continue compiling. Or maybe I do exercise it but I don't care if my user tries something and gets a stack trace.

som-snytt commented:
I guess label comes last because I always write `@deprecated("no go", since="1.0")`.

szeiger commented:
I'm not convinced there are good use cases for such an annotation. The overloaded methods that motivated this PR are already deprecated. The only reason for not removing them completely is cross-compilation. Turning the warnings into errors defeats this purpose. The annotation prevents you from calling code that would otherwise work just fine, which is exactly the opposite of what we want for migration and cross-compiling. As long as I have a working implementation I would never want to use this annotation.

This leaves the cases where you cannot implement a method anymore for some reason.  On the one hand keeping the annotated version around will give you more useful error messages when you compile an old codebase, but on the other hand the unavailable method will show up in scaladocs, IDE/REPL autocomplete and similar features. You need a commitment that the annotation will be supported by all these tools, otherwise it will do more harm than good.

I would prefer more configurable error reporting. The control over these errors and warnings belongs in the hand of the users, not the library authors. Ideally, when I'm cross-compiling, I want to see warnings for all features that were already deprecated in the old version but not necessarily the new version. Then when I upgrade to a newer version I can first drop support for the old one, fix the deprecations and then add the new version.

What I never want to see is gratuitous errors. If the code can work, make it work! I can deal with a large number of deprecations in a huge codebase by fixing them one by one, always compiling and running tests in between to ensure that nothing got broken by the changes. If you turn them into errors, this is no longer possible.

This doesn't mean there can't be different severities of warnings. Other than showing or hiding deprecations based on their version you may want to group and sort them according to versions, projects or other information. But all of this can be done with normal `deprecated` annotations and most of it is not a job for the compiler. IMHO this should be handled at the LSP or BSP level, with the command-line compiler being just another LSP client. Have the compiler emit all kinds of warnings to the server, use a common filtering format (than can be shared between IDEs and command line builds) to perform grouping and filtering in the server, then send the result to the client.

Adding an additional "severity" or "category" string tag to annotations to facilitate this may be useful but first of all we need the filtering infrastructure, even if can only filter by deprecation message, version and source. Adding more information without the tools to use it wouldn't give you any benefit.

som-snytt commented:
Hard deprecation handled by the reporter: https://github.com/scala/scala/pull/7728/commits/9fe4c69d805f4129e9523493c45369cc170c52d0#diff-42e7e257ef752800c72c08bb02e209cdR86

I named it `deprecatedError`, sample https://github.com/scala/scala/pull/7728/commits/9fe4c69d805f4129e9523493c45369cc170c52d0#diff-05a19b860c1823b62b627a31871d792dR26

If `Thread.stop` were `deprecatedError`, I could continue building by de-escalating it. Let's suppose it had a buggy and dangerous implementation and one should never use it. I may still choose not to update my "exit and die" code because it's never actually called. (Maybe `finalize` would be a better example where the line between "supported" and "bad idea" is unclear.)

I don't have a strong opinion, but the use case, "I would just remove this API but I want to give you a nice error message", seems useful.

Maybe just moving it from ref checks to the reporter allays Stefan's suspicions.

eed3si9n commented:
> I'm not convinced there are good use cases for such an annotation.

I've used something like `@compileTimeError` for sbt's old operators `<<=`, `<+=`, and `<++=`.
We didn't want to carry the implementation over to sbt 1, but we also knew there were many old builds around that would benefit from better error message than method not found. So I created a macro just to display an error message - 
 https://github.com/sbt/sbt/pull/2711/commits/01bd4add3cdd96af5e7443dfbe021704ed815f69

> The only reason for not removing them completely is cross-compilation. 

Using it on multi-param `+` may not have been the best first use case? I didn't mean to prevent the normal deprecation cycle. 

> What I never want to see is gratuitous errors. If the code can work, make it work! I can deal with a large number of deprecations in a huge codebase by fixing them one by one, always compiling and running tests in between to ensure that nothing got broken by the changes. If you turn them into errors, this is no longer possible.

Having `@compileTimeError` doesn't preclude going through a normal deprecation cycle first. Eventually the deprecated methods will be dropped. The drop timing is up to the library authors, but migration is up to the users who might skip a few versions (due to high migration cost because of the code size etc).

In other cases, a library may decide to drop a method because the result is incorrect/unsafe and it requires some user action.

> This doesn't mean there can't be different severities of warnings. Other than showing or hiding deprecations based on their version you may want to group and sort them according to versions, projects or other information. But all of this can be done with normal `deprecated` annotations and most of it is not a job for the compiler.

The discussion around restricted/lint/apiStatus is going towards the direction it seems. I don't know if `deprecated` is enough since the existence of `akka.annotation.ApiMayChange`, `akka.annotation.DoNotInherit`, and `akka.annotation.InternalApi` shows that there are some use case that `@deprecated` doesn't exactly fit. The tooling argument is chicken-egg. If there were annotation that the library authors widely adopted to denote the status of the API, I think the tools would be motivated to provide support.

> IMHO this should be handled at the LSP or BSP level, with the command-line compiler being just another LSP client.

In LSP, the compiler plays as the backend to the language server, where the text editor plays the client, similar to browser-web-server relationship. It is conceivable for the language server to send events about this, but it's not exactly a silver bullet since LSP ~= common denominator of what VS Code wants from its backend. Compiler errors and warnings are expressed using [Diagnostic](https://microsoft.github.io/language-server-protocol/specification#diagnostic), but it's not like the editors can configure specific reaction for these. I would imagine Scalafix or some compiler plugin to be a better fit especially if the user would want to enforce the linting rules in CI.

eed3si9n commented:
@som-snytt 

> Hard deprecation handled by the reporter: 9fe4c69#diff-42e7e257ef752800c72c08bb02e209cdR86

Nice.

> Maybe just moving it from ref checks to the reporter allays Stefan's suspicions.

If that's the case I'd gladly hand the torch over to you.

NthPortal commented:
Here's an example of something that really ought to be an error rather than just a deprecation ([doc](https://www.scala-lang.org/api/current/scala/runtime/RichInt.html#round:Int)):

```scala
/** There is no reason to round an `Int`, but this method is provided to avoid accidental loss of precision from a detour through `Float`. */
@deprecated("this is an integer type; there is no reason to round it.  Perhaps you meant to call this on a floating-point value?", "2.11.0")
def round: Int = self
```

som-snytt commented:
`def f(xs: X+)` should mean it takes a nonempty seq, to avoid `f(empty: _*)`.

NthPortal commented:
> `def f(xs: X+)` should mean it takes a nonempty seq, to avoid `f(empty: _*)`.

it seems that would prevent you from passing a `Seq` as varargs

adriaanm commented:
If we're going to get this annotation into 2.13.0, we need to agree on the design and merge this PR this week. I suggest delaying the multi-param infix errors until later -- we can introduce those in 2.13.x. The reporter configuration part can also come later.

For the annotation, I suggest:
```
final class restricted(message: String, label: String, defaultSeverity: String) extends scala.annotation.StaticAnnotation
```

In my mind, "lint" could be a label, but the more generic part is that we're restricting usage of a member (as a generalisation of deprecating). All arguments are stringly typed because we're going for flexibility and user-friendliness. There's little for the type checker to enfore here.

adriaanm commented:
You could encode `akka.annotation.ApiMayChange` as `@restricted("API is evolving", "api-evolving", "error")`, and akka could build with something like `-Wconfig: if (label == "api-evolving" && sourcePackage ~= "akka.*") severity = "ignore"` (the config could be a custom subclass of `Reporter` that implements these rules in Scala).

eed3si9n commented:
### use case 1: library author really wants it to be an error

For this mode, we need something like `@compileTimeError`. As per the name, I'd be happy with `@deprecatedError`.

### use case 2: library author provides some context, and library users decide the severity

This is the use case that generalizes `@deprecated`. That would look like `@restricted` minus `defaultSeverity`.

eed3si9n committed:
deprecatedError annotation

Implements `deprecatedError` annotation, which is a stronger version of `deprecated` annotation. Instead of continuing to compile the code, `deprecatedError` would fail to compile, displaying the message.
The motivation is to display the (hopefully helpful) migration message after the method is gone.

eed3si9n committed:
restricted annotation

This implements `restricted` annotation as a generic form of `deprecated` annotation. `restricted` takes `label` and `defaultSeverity` parameter.

The user can configure how the usages of restricted API are handled by passing in `-Wconfig` config expressions.

A warning config expression may be <label>, <label>:<severity>, or <label>:<prefix>:<severity>

<label> is a String passed into `restricted` annotation.
<severity> must be one of "error", "warning", "info", or "none": default: warning.
<prefix> denotes the prefix of the restricted API. For example, "foo.*" will denote
that this config would apply only to calls under foo package/object.
Note: "*" is simply ignored.

The following configuration would escalate apiMayChange restriction to an error:
    -Wconfig apiMayChange:foo.*:error

eed3si9n commented:
@adriaanm Implemented `@restricted` + `-Wconfig` rules.

adriaanm commented:
Wow, excellent! What do you think about also having `@restricted` subsume `@deprecatedError`?

eed3si9n commented:
> What do you think about also having `@restricted` subsume `@deprecatedError`?

My preference would be to keep `@deprecatedError` as a shorthand, assuming there's no facility in Scala to abstract an annotation with specific arguments.

NthPortal commented on src/library/scala/annotation/restricted.scala:
```diff
@@ -0,0 +1,61 @@
+object RestrictedLabel {
+  final val deprecated   = "deprecated"
+  final val internalOnly = "internalOnly"
+  final val apiMayChange = "apiMayChange"
```
I think a "mistake" label for methods whose use indicates a coding mistake would be nice

SethTisue commented:
@retronym mentioned [JEP 277](https://openjdk.java.net/jeps/277), which is very much in this space; it was implemented in JDK 9. The "Alternatives" section, about alternatives. They strongly considered encoding the reasons for the deprecation, but in the end, decided only to add a single `forRemoval()` method (and `since()`, which our annotation already has):

> The exact reasons for deprecating an API are often too subtle to be expressed as flags or element values in the annotation. It is strongly recommended that the reasons for deprecating an API be described in that API's documentation comments. In addition, it is also recommended that potential replacement APIs be discussed and linked from the documentation

> A string value has been proposed as a detail code. This appears to provide more flexibility, but it also introduces problems with weak typing and namespace conflicts, possibly leading to undetected errors

> Previous versions of this proposal included a variety of "reason" codes including UNSPECIFIED, DANGEROUS, OBSOLETE, SUPERSEDED, UNIMPLEMENTED, and EXPERIMENTAL. These attempted to encode the reason for which an API was deprecated, the risks of using it, and also whether a replacement API is available. In practice, all of this information is too subjective be encoded as values in an annotation. Instead, this information should be described in the Javadoc documentation comment.

adriaanm commented:
In the mentioned team meeting I fought pretty hard for this PR, but couldn't convince the others. So, closing, with regret. In summary, the conclusion was that we need to flesh out the design space more and probably submit a SIP.

som-snytt commented:
Also forRemoval https://github.com/scala/scala/pull/7714

There is a certain wisdom in leveraging the work already done by the Java community.

However, the optional label in the other PR is for convenient suppression, not for explanatory purposes, so kinda orthogonal.
